File: src/common/py/integration/authorization/strategies/oauth2_strategy.py
import json
import time
import typing
from dataclasses import dataclass

import requests
from dataclasses_json import dataclass_json

from .authorization_strategy import AuthorizationStrategy
from .oauth2_types import OAuth2Token, OAuth2AuthorizationRequestData, OAuth2TokenData
from ....component import BackendComponent
from ....data.entities.authorization import AuthorizationToken
from ....data.entities.user import UserID
from ....services import Service


@dataclass_json
@dataclass(frozen=True, kw_only=True)
class OAuth2Configuration:
    """
    The OAuth2 strategy configuration.
    """


class OAuth2Strategy(AuthorizationStrategy):
    """
    OAuth2 authorization strategy.
    """

    Strategy: str = "oauth2"

    def __init__(
        self, comp: BackendComponent, svc: Service, config: OAuth2Configuration
    ):
        super().__init__(comp, svc, OAuth2Strategy.Strategy)

        self._config = config

    def request_authorization(
        self, user_id: UserID, auth_id: str, request_data: typing.Any
    ) -> AuthorizationToken:
        oauth2_data = self._get_request_data(request_data)
        client_secret = self._get_client_secret(auth_id)

        response = requests.post(
            oauth2_data.token_endpoint,
            data={
                "grant_type": "authorization_code",
                "client_id": oauth2_data.client_id,
                "client_secret": client_secret,
                "code": oauth2_data.auth_code,
                "redirect_uri": oauth2_data.redirect_url,
            },
        )

        if response.status_code == 200:
            resp_data = response.json()
            try:
                self._verify_oauth2_token_data(resp_data)

                return AuthorizationToken(
                    user_id=user_id,
                    auth_id=auth_id,
                    expiration_timestamp=(
                        time.time() + resp_data["expires_in"]
                        if "expires_in" in resp_data
                        else 0
                    ),
                    strategy=self.strategy,
                    token=OAuth2Token(
                        access_token=resp_data["access_token"],
                        token_type=resp_data["token_type"],
                        refresh_token=resp_data["refresh_token"],
                    ),
                    data=OAuth2TokenData(
                        token_endpoint=oauth2_data.token_endpoint,
                        client_id=oauth2_data.client_id,
                    ),
                )
            except Exception as exc:  # pylint: disable=broad-exception-caught
                raise RuntimeError(f"Invalid OAuth2 token received: {exc}")
        else:
            raise RuntimeError(
                f"Unable to request an access token: {response.text}"
            )  # TODO: Proper err

    def _get_request_data(
        self, request_data: typing.Any
    ) -> OAuth2AuthorizationRequestData:
        oauth2_data = OAuth2AuthorizationRequestData.from_dict(request_data)

        # Verify the request data
        if oauth2_data.token_endpoint == "":
            raise RuntimeError("Missing token endpoint")
        if oauth2_data.client_id == "":
            raise RuntimeError("Missing client ID")
        if oauth2_data.auth_code == "":
            raise RuntimeError("Missing authentication code")
        if oauth2_data.redirect_url == "":
            raise RuntimeError("Missing redirection URL")

        return oauth2_data

    def _get_client_secret(self, auth_id: str) -> str:
        client_secret = self._get_config_value(f"secrets.{auth_id}", "")

        # Verify the secret
        if client_secret == "":
            raise RuntimeError(f"Missing OAuth2 client secret for {auth_id}")

        return client_secret

    def _verify_oauth2_token_data(self, data: typing.Dict[str, typing.Any]) -> None:
        if "access_token" not in data or data["access_token"] == "":
            raise RuntimeError("Missing access token")
        if "token_type" not in data or data["token_type"] == "":
            raise RuntimeError("Missing token type")
        if "refresh_token" not in data or data["refresh_token"] == "":
            raise RuntimeError("Missing refresh token")


def create_oauth2_strategy(
    comp: BackendComponent, svc: Service, config: typing.Any
) -> OAuth2Strategy:
    """
    Creates a new OAuth2 strategy instance, automatically configuring it.

    Args:
        comp: The main component.
        svc: The service to use for message sending.
        config: The strategy configuration.

    Returns:
        The newly created strategy.
    """
    if not isinstance(config, OAuth2Configuration):
        raise RuntimeError("Invalid configuration passed for OAuth2")

    oauth2_config = typing.cast(OAuth2Configuration, config)

    return OAuth2Strategy(comp, svc, oauth2_config)

File: src/common/web/integration/authorization/strategies/OAuth2Strategy.ts
import { string } from "yup";
import { WebComponent } from "../../../component/WebComponent";
import { AuthorizationState } from "../../../data/entities/authorization/AuthorizationState";
import { Service } from "../../../services/Service";
import { getURLQueryParam } from "../../../utils/URLUtils";
import { type AuthorizationRequestData, AuthorizationStrategy } from "./AuthorizationStrategy";

/**
 * The OAuth2 strategy configuration.
 */
export interface OAuth2Configuration {
    server: {
        endpoints: {
            authorization: string;
            token: string;
        };
    };
    client: {
        clientID: string;
        redirectURL: string;
        embedded: boolean;
    };
}

/**
 * OAuth2 authorization request data.
 */
export interface OAuth2AuthorizationRequestData {
    auth_code: string;
    token_endpoint: string;
}

/**
 * OAuth2 authorization strategy.
 */
export class OAuth2Strategy extends AuthorizationStrategy {
    public static readonly Strategy = "oauth2";

    private readonly _config: OAuth2Configuration;

    public constructor(comp: WebComponent, svc: Service, config: OAuth2Configuration) {
        super(comp, svc, OAuth2Strategy.Strategy, config.client.embedded);

        this._config = config;
    }

    protected initiateRequest(fingerprint: string): void {
        const url = new URL(this._config.server.endpoints.authorization);
        url.searchParams.set("response_type", "code");
        url.searchParams.set("client_id", this._config.client.clientID);
        url.searchParams.set("redirect_uri", this._config.client.redirectURL);
        url.searchParams.set("state", fingerprint);
        this.redirect(url.toString());
    }

    protected getRequestData(): AuthorizationRequestData {
        const authCode = getURLQueryParam("auth:code");
        const fingerprint = getURLQueryParam("auth:fingerprint");
        if (!authCode || !fingerprint) {
            throw new Error("No authentication information provided");
        }

        return {
            data: {
                auth_code: authCode,
                token_endpoint: this._config.server.endpoints.token,
            } as OAuth2AuthorizationRequestData,
            fingerprint: fingerprint,
        } as AuthorizationRequestData;
    }

    protected finishRequest(authState: AuthorizationState): void {
        if (authState == AuthorizationState.Authorized) {
            this.redirect(this._config.client.redirectURL);
        }
    }
}

/**
 * Creates a new OAuth2 strategy instance, automatically configuring it.
 *
 * @param comp - The main component.
 * @param svc - The service to use for message sending.
 * @param config - The strategy configuration.
 *
 * @returns - The newly created strategy.
 */
export function createOAuth2Strategy(comp: WebComponent, svc: Service, config: Record<string, any>): OAuth2Strategy {
    const oauth2Config = config as OAuth2Configuration;

    // Verify the passed configuration
    if (!oauth2Config.server?.endpoints?.authorization) {
        throw new Error("Missing authorization endpoint");
    }
    if (!oauth2Config.server?.endpoints?.token) {
        throw new Error("Missing token endpoint");
    }

    if (!oauth2Config.client?.clientID) {
        throw new Error("Missing client ID");
    }
    if (!oauth2Config.client?.redirectURL) {
        throw new Error("Missing redirection URL");
    }

    return new OAuth2Strategy(comp, svc, oauth2Config);
}
